pkg/p2p/libp2p/internal/breaker/breaker.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package breaker
+
+import (
+	"errors"
+	"sync"
+	"time"
+)
+
+var (
+	_ Interface = (*breaker)(nil)
+
+	// timeNow is used to deterministically mock time.Now() in tests
+	timeNow = time.Now
+
+	// ErrClosed is the special error type that indicates that breaker is closed and that is not executing functions at the moment.
+	ErrClosed = errors.New("breaker closed")
+)
+
+type Interface interface {
+	Execute(f func() error) error
+}
+
+type breaker struct {
+	limit                int
+	consFailedCalls      int
+	firstFailedTimestamp time.Time
+	closedTimestamp      time.Time
+	backoff              time.Duration
+	maxBackoff           time.Duration
+	failInterval         time.Duration
+	mtx                  sync.Mutex
+}
+
+type Options struct {
+	Limit        int
+	FailInterval time.Duration
+	StartBackoff time.Duration
+	MaxBackoff   time.Duration
+}
+
+func NewBreaker(o Options) Interface {
+	breaker := &breaker{
+		limit:        o.Limit,
+		backoff:      o.StartBackoff,
+		maxBackoff:   o.MaxBackoff,
+		failInterval: o.FailInterval,
+	}
+
+	if o.Limit == 0 {
+		breaker.limit = 100
+	}
+
+	if o.FailInterval == 0 {
+		breaker.failInterval = 30 * time.Minute
+	}
+
+	if o.MaxBackoff == 0 {
+		breaker.maxBackoff = time.Hour
+	}
+
+	if o.StartBackoff == 0 {
+		breaker.backoff = 2 * time.Minute
+	}
+
+	return breaker
+}
+
+// Execute runs f() if the limit number of consecutive failed calls is not reached within fail interval.
+// f() call is not locked so it can still be executed concurently.
+// Returns `ErrClosed` if the limit is reached or f() result otherwise.
+func (b *breaker) Execute(f func() error) error {
```

Comment:
Since this type is not exported, this comment could be set on the interface method or as in the package doc, so that it will be visible in godoc.